atomic_reactor/plugins/post_compare_components.py

```diff
@@ -0,0 +1,90 @@
+
+
+class CompareComponentsPlugin(PostBuildPlugin):
+    """
+    Compare components from each worker build and verify the same version was
+    on each worker.
+    """
+
+    key = PLUGIN_COMPARE_COMPONENTS_KEY
+    is_allowed_to_fail = False
+
+    def rpm_compare(self, a, b):
+        """
+        Compare rpm component version 'a' with 'b'.
+        'name' is implied equal as it is the key() lookup
+        """
+        self.log.debug("%s: version %s == %s" % (a['name'], a['version'], b['version']))
+        if a['version'] == b['version'] and \
+           a['release'] == b['release'] and \
+           a['signature'] == b['signature']:
+            return
+
+        self.log.debug("Comparing %s to %s" % (a, b))
+        raise PluginFailedException("Version mismatch for package %s" % a['name'])
+
+    def get_component_list(self, worker_metadatas):
+        """
+        Find the worker build's component list.
+        The strategy is to find the 'output' key in the metadata and then
+        search the array for a 'docker-image' type.  The 'log' types are not
+        interesting.
+
+        :return: array of components
+        """
+        comp_list = []
+        for platform in sorted(worker_metadatas.keys()):
+            for instance in worker_metadatas[platform]['output']:
+                if instance['type'] == 'docker-image':
+                    if 'components' not in instance:
+                        self.log.warn("Missing 'components' key in 'output' metadata instance: %s",
+                                      instance)
+                        continue
+
+                    comp_list.append(instance['components'])
+
+        return comp_list
+
+    def run(self):
+        """
+        Run the plugin.
+        """
+
+        worker_metadatas = self.workflow.postbuild_results.get(PLUGIN_FETCH_WORKER_METADATA_KEY)
+        comp_list = self.get_component_list(worker_metadatas)
+
+        # master compare list
+        master_comp = {}
+
+        # The basic strategy is to start with empty lists and add new component
+        # versions as we find them.  Upon next iteration, we should notice
+        # duplicates and be able to compare them.  If the match fails, we raise
+        # an exception.  If the component name does not exist, assume it was an
+        # arch dependency, add it to list and continue.  By the time we get to
+        # the last arch, we should have every possible component in the master
+        # list to compare with.
+
+        # Keep everything separated by component type
+        for components in comp_list:
+            for component in components:
+                t = component['type']
+                name = component['name']
+
+                if name not in master_comp:
+                    master_comp[name] = component
+                    continue
+
+                if t == 'rpm':
+                    self.rpm_compare(master_comp[name], component)
```

Comment:
Do we want to collect all errors then raise exception? Might be a better user experience.